internal/source/source.go

```diff
@@ -0,0 +1,67 @@
+package source
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"sync"
+
+	rukpakv1alpha1 "github.com/operator-framework/rukpak/api/v1alpha1"
+)
+
+type Source interface {
+	SourceContent(context.Context, *rukpakv1alpha1.Bundle) (*Result, error)
+}
+
+type Result struct {
+	Bundle         fs.FS
+	ResolvedSource *rukpakv1alpha1.BundleSource
+	State          State
+	Message        string
+}
+
+type State string
+
+const (
+	StatePending   State = "Pending"
+	StateUnpacking State = "Unpacking"
+	StateUnpacked  State = "Unpacked"
+)
+
+type Sourcer struct {
+	sources map[rukpakv1alpha1.BundleSourceType]Source
+	m       sync.RWMutex
+}
+
+var (
+	ErrSourceAlreadyAdded = errors.New("source type already added")
+	ErrSourceNotSupported = errors.New("source type not supported")
+)
+
+func NewSourcer() *Sourcer {
+	return &Sourcer{sources: map[rukpakv1alpha1.BundleSourceType]Source{}}
+}
+
+func (s *Sourcer) Add(typ rukpakv1alpha1.BundleSourceType, source Source) error {
+	s.m.RLock()
+	_, ok := s.sources[typ]
+	s.m.RUnlock()
+	if ok {
+		return ErrSourceAlreadyAdded
+	}
+
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.sources[typ] = source
+	return nil
+}
+
+func (s *Sourcer) For(typ rukpakv1alpha1.BundleSourceType) (Source, error) {
```

Comment:
Just to clarify my understanding: we're trying to check whether the type parameter we'd passed is a valid, supported Bundle source type for this provisioner. Is that correct? Is there a better name for this method?

Reply:
I think I'm slowly coming back around to this naming after the first pass through these changes. I'd still like to see a comment that helps explain the functionality so it's a bit easier to reason about when interacting with the codebase, but this isn't blocking anyways.

Reply:
It's effectively a map lookup. "Get the `Source` for this type". `Has` wouldn't make sense (as that would intuitively return a `bool`). `Get` could make sense. I tend to shy away from functions called `Get` since they're so generic sounding.

Seeing something like `src, err := sourcer.For(rukpakv1alpha1.BundleSourceTypeImage)` reads pretty nicely to me.

Regardless, +1 on GoDoc.

Reply:
Yep I think I'm fine with the current naming after a couple of read through(s) with these changes, and I think my first read would've been easier if we integrated godocs for that method but this isn't blocking anyways so feel free to resolve.

Reply:
I changed this up one more time, such that `Sourcer` is now an unexported struct that implements the `Source` interface. So now the controller just immediately calls `Unpack` on it's `Source` rather than getting the underlying source and then calling `Unpack`. I think this is a bit cleaner.